design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+to extend the capabilities of KSQL without the need to change the language in a breaking way.
+
+
+
+## Public APIS
+
+We propose the following language changes:
+
+* introduce `TABLE(<stream>)` operator: takes a STREAM as input, and changes its semantic interpretation from facts to updates; it returns a TABLE
+  * note: the returned TABLE is not an RDBMS style table, but rather a MV
+  * note: it's not specified if the returned TABLE is materialized into a state store or represented as a changelog
+    (it's up the the KSQL optimizer or KS runtime to make a materialization decision)
+  * the `TABLE()` operator consumes it's input STREAM "from beginning"
+  * we could extend `TABLE()` to also accept a TABLE as input if it makes the language smoother;
+    it's obvious that `TABLE(<table>)` would be idempotent, i.e., `TABLE(t) = t`
+* introduce `STREAM(<table>)` operator: takes a TABLE as input, and returns a STREAM that contains a record for each update to the TABLE
+  * note: the returned STREAM is a fact stream, and would be processed with corresponding STREAM semantics (it's semantically not a changelog stream)
+  * note: the input TABLE can either be an RDBMS style table of a MV
+  * by default, the output STREAM contains one record for each input table record (full table scan on the table generation when the query was issues)
+    plus a record for each input table update
+  * we could extend `STREAM()` to also accept a STREAM as input if it makes the language smoother;
+  it's obvious that `STREAM(<stream>)` would be idempotent, i.e., `STREAM(s) = s`
+* Introduce an `AS OF <timestamp>` clause for RDBMS style queries against tables
+* Introduce `CREATE MATERIALIZED VIEW` statement
+* Deprecate `CREATE TABLE` statement (in favor of the new `CREATE MATERIALIZED VIEW` statement) and eventually remove it
+* Deprecate `EMIT CHANGES` clause
+
+Semantic changes:
+
+* An aggregation query over a stream returns a stream (and not a table)
+
+Naming:
+
+* Drop the terms `pull query` and `push query`
+* Use the terms `table query` and `stream query`
+  (this will also allow to use the phrase "query a stream" and "query a table" in a straight forward way)
+
+
+
+## Design
+
+We introduce new syntax step by step and illustrate it by examples. We also explain the corresponding semantics.
+
+
+**Example 0: create persistent streams/tables**
+
+Syntax is up for discussion (it's possible to merge some statements) – we put out multiple ideas for brainstorming.
+
+```sql
+-- no change to existing `CREATE STREAM` statements
+CREATE STREAM <name> FROM <topic>
+
+CREATE STREAM <name> AS <query>   --the query must return a STREAM
+
+
+
+-- new syntax that replaces existing `CREATE TABLE`
+CREATE MATERIALIZED VIEW <name> FROM <topic>
+
+CREATE MATERIALIZED VIEW <name> FROM <stream>   ---the <stream> must be a registered stream from the catalog
+
+CREATE MATERIALIZED VIEW <name> AS <query>   --the query must return a STREAM
```

Comment:
Why not create a MV from another MV or from a table?

Reply:
The `<query>` can either be a stream query or a table query so what you ask is included. Does this make sense?